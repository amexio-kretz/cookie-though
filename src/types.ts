export type CookiePreference = Pick<Policy, 'id'> & {
  isEnabled: boolean;
};

export interface CookiePreferences {
  isCustomised: boolean;
  cookieOptions: CookiePreference[];
}

export type CookieOption = Policy & {
  isEnabled: boolean;
};

export interface CookieOptions extends CookiePreferences {
  cookieOptions: CookieOption[];
}
export interface Policy {
  id: string;
  label: string;
  description: string;
  category: 'essential' | 'social' | 'personalisation' | 'marketing' | 'statistics' | 'functional';
}

export interface Config {
  policies: Policy[];
  permissionLabels: {
    accept: string;
    acceptAll: string;
    decline: string;
  };
  cookiePreferenceKey?: string;
  header?: {
    title?: string;
    subTitle?: string;
    description?: string;
  };
  cookiePolicy?: {
    url: string;
    label: string;
  };
}
