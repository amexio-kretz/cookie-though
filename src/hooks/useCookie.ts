import { CookieOption } from '../types';
import { CookiePreferences } from '../components/app';

export const COOKIE_PREFERENCES_KEY = 'cookie-preferences';

interface Options {
  cookieOptions: CookieOption[];
  cookiePreferenceKey?: string;
}

export const getCookie = (cookieKey: string): CookiePreferences | undefined => {
  const rawCookies = decodeURIComponent(document.cookie).split(';');
  return rawCookies.reduce((cookiePreferences, cookie) => {
    if (cookiePreferences) return cookiePreferences;

    const [key, value] = cookie.split('=');
    if (key === cookieKey) return JSON.parse(value);
  }, undefined);
};

export const getNextYear = () => {
  const now = new Date();
  return new Date(now.setFullYear(now.getFullYear() + 1));
};

const policiesHaveChanged = (
  policyIds: CookieOption['id'][],
  preferenceIds: CookieOption['id'][],
) => {
  return policyIds.some(id => !policyIds.includes(id)) || policyIds.length !== preferenceIds.length;
};

const mergePolicies = (policies: CookieOption[], preferences: CookieOption[]): CookieOption[] => {
  return policies.map(policy => {
    const isEnabled = preferences.reduce((isPreferenceEnabled, preference) => {
      return preference.id === policy.id ? preference.isEnabled : isPreferenceEnabled;
    }, false);

    return { ...policy, isEnabled };
  });
};

const useCookie = ({ cookieOptions, cookiePreferenceKey = COOKIE_PREFERENCES_KEY }: Options) => {
  const preferences: CookiePreferences = {
    cookieOptions,
    isCustomised: false,
  };

  const setCookiePreferences = (cookiePreferences: CookiePreferences) => {
    const expires = getNextYear();
    document.cookie = `${cookiePreferenceKey}=${JSON.stringify(
      cookiePreferences,
    )}; expires=${expires}; path=/`;

    return cookiePreferences;
  };

  const getCookiePreferences = () => {
    const cookiePreferences = getCookie(cookiePreferenceKey);

    if (!cookiePreferences) return setCookiePreferences(preferences);

    const policyIds = preferences.cookieOptions.map(cookieOption => cookieOption.id);
    const preferenceIds = cookiePreferences.cookieOptions.map(cookieOption => cookieOption.id);
    if (!policiesHaveChanged(policyIds, preferenceIds)) return cookiePreferences;

    const cookieOptions = mergePolicies(preferences.cookieOptions, cookiePreferences.cookieOptions);

    return { cookieOptions, isCustomised: false };
  };

  return { getCookiePreferences, setCookiePreferences };
};

export default useCookie;
