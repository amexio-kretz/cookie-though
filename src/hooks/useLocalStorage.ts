import { CookieOption } from '../types';
import { CookiePreferences } from './../components/app';

export const COOKIE_PREFERENCES_KEY = 'cookie-preferences';

const useLocalStorage = (cookieOptions: CookieOption[]) => {
  const DEFAULT_COOKIE_PREFERENCES: CookiePreferences = {
    cookieOptions,
    isCustomised: false,
  };

  const setCookiePreferences = (cookiePreferences: CookiePreferences) => {
    localStorage.setItem(COOKIE_PREFERENCES_KEY, JSON.stringify(cookiePreferences));
    return cookiePreferences;
  };

  const getCookiePreferences = () => {
    const rawCookiePreferences = localStorage.getItem(COOKIE_PREFERENCES_KEY);
    return rawCookiePreferences
      ? (JSON.parse(rawCookiePreferences) as CookiePreferences)
      : setCookiePreferences(DEFAULT_COOKIE_PREFERENCES);
  };

  return { getCookiePreferences, setCookiePreferences };
};

export default useLocalStorage;
