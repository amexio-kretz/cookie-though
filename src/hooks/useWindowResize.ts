import { useLayoutEffect } from 'preact/hooks';
import throttle from 'lodash.throttle';

const useWindowResize = (callback: () => void) => {
  useLayoutEffect(() => {
    const throttledCallback = throttle(callback, 100);
    window.addEventListener('resize', throttledCallback);

    return () => window.removeEventListener('resize', throttledCallback);
  }, [callback]);
};

export default useWindowResize;
