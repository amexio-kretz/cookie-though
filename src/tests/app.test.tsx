import { h } from 'preact';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import mockPolicies from './__mocks__/policies';
import { COOKIE_PREFERENCES_KEY } from '../hooks/useCookie';
import mockConfig, { dutchMockConfig } from './__mocks__/config';
import CookieThough, { App } from '../components/app';
import clearCookies from './utils/clearCookies';
import { CookiePreferences } from '../types';

describe('Cookie Though', () => {
  afterEach(() => {
    clearCookies();
    document.getElementsByTagName('html')[0].innerHTML = '';
  });

  describe('init function', () => {
    beforeEach(() => {
      const manageCookiesElement = document.createElement('button');
      manageCookiesElement.id = 'manage-cookie-though';
      document.body.append(manageCookiesElement);
    });

    it('can render the app based on the initCookieThough function', () => {
      CookieThough.init({ ...mockConfig });

      expect(document.querySelector('.cookie-though')).toBeDefined();
    });

    it('can switch the config', () => {
      CookieThough.init({ ...mockConfig });
      const shadowRoot = document.querySelector('.cookie-though')?.shadowRoot as ShadowRoot;

      let cookiePolicyLink = shadowRoot.querySelector('a');
      expect(cookiePolicyLink?.text).toEqual(mockConfig.cookiePolicy.label);

      CookieThough.init({ ...dutchMockConfig });
      expect(document.getElementsByClassName('cookie-though').length).toEqual(1);
      cookiePolicyLink = shadowRoot.querySelector('a');
      expect(cookiePolicyLink?.text).toEqual(dutchMockConfig.cookiePolicy.label);
    });

    it('can hide the cookie wall with the setVisible function', () => {
      const { setVisible } = CookieThough.init({
        ...mockConfig,
      });

      expect(document.querySelector('.cookie-though')).toBeDefined();
      expect(document.querySelector('.visible')).toBeDefined();
      setVisible(false);
      expect(document.querySelector('.visible')).toBeNull();
    });

    it('will call the onPreferencesChanged callback when the preferences are updated', () => {
      const onPreferencesChanged = jest.fn((preferences: CookiePreferences) => {
        expect(preferences).toEqual({
          cookieOptions: [
            {
              id: 'essential',
              isEnabled: true,
            },
            {
              id: 'functional',
              isEnabled: true,
            },
            {
              id: 'analytics',
              isEnabled: true,
            },
            {
              id: 'marketing',
              isEnabled: true,
            },
          ],
          isCustomised: true,
        });
      });
      CookieThough.init({ ...mockConfig });
      const shadowRoot = document.querySelector('.cookie-though')?.shadowRoot as ShadowRoot;

      const acceptAllButton = Array.from(shadowRoot.querySelectorAll('button')).find(
        button => button.textContent === 'Accept all',
      );
      CookieThough.listen(onPreferencesChanged);
      acceptAllButton?.click();
      expect(onPreferencesChanged).toBeCalledTimes(1);
    });
  });

  describe('without user preferences stored in local storage', () => {
    it('should show the cookie wall', () => {
      const setVisible = jest.fn((value: boolean) => {
        expect(value).toBeTruthy();
      });
      mount(
        <div className="cookie-though">
          <App
            policies={mockConfig.policies}
            permissionLabels={mockConfig.permissionLabels}
            setVisible={setVisible}
          />
        </div>,
        { attachTo: document.body },
      );

      expect(setVisible).toBeCalled();
    });
  });

  describe('with user preferences stored in a cookie', () => {
    const DEFAULT_COOKIE_PREFERENCES: CookiePreferences = {
      isCustomised: false,
      cookieOptions: mockConfig.policies.map(policy => ({ ...policy, isEnabled: false })),
    };

    it('should not show the cookie wall', () => {
      document.cookie = `${COOKIE_PREFERENCES_KEY}=${JSON.stringify({
        ...DEFAULT_COOKIE_PREFERENCES,
        isCustomised: true,
      })}`;
      const setVisible = function () {
        throw new Error('should not be called');
      };
      mount(
        <div className="cookie-though">
          <App
            policies={mockPolicies}
            permissionLabels={mockConfig.permissionLabels}
            setVisible={setVisible}
          />
        </div>,
        { attachTo: document.body },
      );
    });

    it("should show the cookie wall if the cookie preferences aren't customised", () => {
      const setVisible = jest.fn((value: boolean) => {
        expect(value).toBeTruthy();
      });

      mount(
        <div className="cookie-though">
          <App
            policies={mockPolicies}
            permissionLabels={mockConfig.permissionLabels}
            setVisible={setVisible}
          />
        </div>,
        { attachTo: document.body },
      );

      expect(setVisible).toBeCalled();
    });
  });

  it('should render properly', () => {
    const wrapper = shallow(
      <body>
        <button id="manage-cookie-though"></button>
        <div className="cookie-though">
          <App
            policies={mockPolicies}
            permissionLabels={mockConfig.permissionLabels}
            setVisible={() => {
              return;
            }}
          />
        </div>
      </body>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
