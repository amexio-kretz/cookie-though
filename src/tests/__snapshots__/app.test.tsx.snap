// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cookie Though should render properly 1`] = `
<body>
  <button
    id="manage-cookie-though"
  />
  <div
    className="cookie-though"
  >
    <App
      permissionLabels={
        Object {
          "accept": "Accept",
          "acceptAll": "Accept all",
          "decline": "Decline",
        }
      }
      policies={
        Array [
          Object {
            "category": "essential",
            "description": "Always on",
            "id": "essential",
            "label": "Strictly necessary cookies",
          },
          Object {
            "category": "functional",
            "description": "We’ll remember the basics such as language.",
            "id": "functional",
            "label": "Functional cookies",
          },
          Object {
            "category": "statistics",
            "description": "We’ll know where we should improve your experience.",
            "id": "analytics",
            "label": "Analytics",
          },
          Object {
            "category": "marketing",
            "description": "We’ll only show you content that interests you.",
            "id": "marketing",
            "label": "Marketing",
          },
        ]
      }
      setVisible={[Function]}
    />
  </div>
</body>
`;
